Rename OpenID Connect API urls to be under /api

zachmargolis opened the pull request:
**Why**: For consistency with other API urls, like /api/saml

---

This is a proposal in the form of a PR, that we move the canonical API URLs for OpenID connect under the `/api` namespace.

Clients that use auto-discovery will not have to change, and if we merged this, I would follow up with the developers we know about and ask them to migrate their clients. Then, we could consider removing the old URLs completely.

monfresh commented on config/application.rb:
```diff
@@ -30,6 +30,14 @@ class Application < Rails::Application
+
+        # legacy URLs
         resource '/openid_connect/certs', headers: :any, methods: [:get]
         resource '/openid_connect/token',
```
Should these endpoints without the `api` prefix be removed?

zachmargolis replied:
I wanted to keep them for backwards compatibility purposes

monfresh commented on config/routes.rb:
```diff
@@ -79,11 +82,10 @@
+  get '/openid_connect/certs' => 'openid_connect/certs#index', as: :old_openid_connect_certs
+  post '/openid_connect/token' => 'openid_connect/token#create', as: :old_openid_connect_token
+  get '/openid_connect/userinfo' => 'openid_connect/user_info#show',
+      as: :old_openid_connect_userinfo
```
What is the purpose of making this change backwards compatible? Since we haven't launched yet, can't we tell any clients that might be using the old endpoints to use the new endpoints?

zachmargolis replied:
The purpose of keeping these backwards compatible would be easier communication with the teams currently developing against the API, and a little bit of leeway in terms of deploy schedules for us and them to get them to switch?

zachmargolis commented on app/forms/openid_connect_token_form.rb:
```diff
@@ -87,16 +87,25 @@ def validate_client_assertion
 
+  def validate_aud_claim(payload)
+    normalized_aud = payload['aud'].to_s.chomp('/')
+    return if [old_openid_connect_token_url, api_openid_connect_token_url].include?(normalized_aud)
```
@jessieay are these the somewhat unrelated changes you're referring to? The main part was accepting either the old or the new token URL for the `aud` value, and then I figured as long as I was here I'd normalize the trailing slash. If we're going to be backwards compatible, I'd definitely like to keep this change

jessieay replied:
Oh I see, I didn't realize that this change was required for the new URLs. If that is the case, sounds good to keep!

zachmargolis commented:
The backwards compatibility is short-term, I'd like to remove the legacy endpoints, but supporting both at once helps us ensure a smooth transition.

I've added specs for the legacy routes, which also helped me discover we we're supporting the OPTIONS request correctly, so I've added support for that too.
